src/emmy/expression/render.cljc

```diff
@@ -661,27 +644,73 @@
-      (let [x      (v/freeze x)
-            params (set/difference (x/variables-in x) operators-known)
-            ordered-params (if (fn? parameter-order)
-                             (parameter-order params)
-                             parameter-order)
-            callback (fn [new-expression new-vars]
-                       (doseq [[var val] new-vars]
-                         (print "  var ")
-                         (print (str var " = "))
-                         (print (R val))
-                         (print ";\n"))
-                       (print "  return ")
-                       (print (R new-expression))
-                       (print ";\n}"))
-            opts {:deterministic?   deterministic?
-                  :symbol-generator symbol-generator}]
-        (with-out-str
-          (print "function(")
-          (print (s/join ", " ordered-params))
-          (print ") {\n")
-          (compile/extract-common-subexpressions x callback opts))))))
+    (fn [x opts]
+      (let [x              (v/freeze x)
+            params         (set/difference (x/variables-in x) operators-known)
+            callback       (fn [new-expression new-vars]
+                             {:params params
+                              :vars   (into [] (map (fn [[k v]] [k (R v)])) new-vars)
+                              :value  (R new-expression)})
+            opts           (merge {:symbol-generator (make-symbol-generator "_")} opts)]
+        (extract-common-subexpressions x callback opts)))))
+
+(defn ->JavaScript
+  "Convert the given expression to a string representation of a
+  JavaScript function.
+
+  Parameters to the function will be extracted from the symbols in the expression.
+  Common subexpression elimination will be performed and auxiliary variables
+  will be bound in the body of the function; the names of these symbols are
+  `_0001, ...`.
+
+  If `:parameter-order` is specified, it is used to determine function parameter
+  order in one of two ways:
+
+  If it is set to a function, that function will be called on the sequence of
+  parameters and is expected to return the parameters in the desired sequence.
+
+  Otherwise, it is interpreted as the sequence of parameters itself. If not
+  specified, the default behavior is `sort`.
+
+  If `:deterministic? true` is supplied, the function will assign variables by
+  sorting the string representations of each term before assignment. Otherwise,
+  the nondeterministic order of hash maps inside this function won't guarantee a
+  consistent variable naming convention in the returned function. For tests, set
+  `:deterministic? true`."
+  [x & {:keys [parameter-order]
+        :or   {parameter-order sort}
+        :as opts}]
+  (let [{:keys [params vars value]} (->JavaScript* x opts)
+        ordered-params (if (fn? parameter-order)
+                         (parameter-order params)
+                         parameter-order)]
+    (with-out-str
```

Comment:
I think this approach of with-out-str is fine for now, but I did hit some bugs a while ago where a println somewhere deep inside of emmy was triggering and getting into the string. No need to change it now, but I wanted to get that down somewhere!

Reply:
Good point. I heard a slight discordant note in the soundtrack as I used this feature and now repent of it. I'll fix it